types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,645 @@
+
+	wlr_log(WLR_DEBUG, "Offering output %s", output->name);
+
+	struct wlr_drm_lease_device_v1 *device = NULL, *tmp_device;
+	wl_list_for_each(tmp_device, &manager->devices, link) {
+		if (tmp_device->backend == output->backend) {
+			device = tmp_device;
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"offered output");
+		return false;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector;
+	wl_list_for_each(connector, &device->connectors, link) {
+		if (connector->output == output) {
+			wlr_log(WLR_ERROR, "Output %s has already been offered",
+					output->name);
+			return false;
+		}
+	}
+
+	connector = calloc(1, sizeof(struct wlr_drm_lease_connector_v1));
+	if (!connector) {
+		wlr_log(WLR_ERROR, "Failed to allocatr wlr_drm_lease_connector_v1");
+		return false;
+	}
+
+	connector->output = output;
+	connector->device = device;
+
+	wl_signal_add(&output->events.destroy, &connector->destroy);
+	connector->destroy.notify = handle_output_destroy;
+
+	wl_list_init(&connector->resources);
+	wl_list_insert(&device->connectors, &connector->link);
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		drm_lease_connector_v1_send_to_client(connector, resource);
+	}
+
+	return true;
+}
+
+
+void wlr_drm_lease_v1_manager_withdraw_output(
+	struct wlr_drm_lease_v1_manager *manager, struct wlr_output *output) {
+	assert(manager && output);
+
+	wlr_log(WLR_DEBUG, "Withdrawing output %s", output->name);
+
+	struct wlr_drm_connector *drm_connector =
+			get_drm_connector_from_output(output);
+
+	struct wlr_drm_lease_device_v1 *device = NULL, *tmp_device;
+	wl_list_for_each(tmp_device, &manager->devices, link) {
+		struct wlr_drm_backend *backend = get_drm_backend_from_backend(
+				tmp_device->backend);
+		if (backend == drm_connector->backend) {
+			device = tmp_device;
+			break;
+		}
+	}
+
+	if (!device) {
+		wlr_log(WLR_ERROR, "No wlr_drm_lease_device_v1 associated with the "
+				"given output");
+		return;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector = NULL, *tmp_conn;
+	wl_list_for_each(tmp_conn, &device->connectors, link) {
+		if (tmp_conn->output == output) {
+			connector = tmp_conn;
+			break;
+		}
+	}
+
+	if (!connector) {
+		wlr_log(WLR_DEBUG, "Output has not been offered");
+		return;
+	}
+
+	if (connector->active_lease) {
+		wlr_drm_lease_v1_revoke(connector->active_lease);
+	}
+}
+
+static void drm_lease_device_v1_destroy(
+		struct wlr_drm_lease_device_v1 *device) {
+	struct wlr_drm_backend *backend =
+			get_drm_backend_from_backend(device->backend);
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_device_v1 for %s",
+			backend->name);
+
+	if (!device) {
+		return;
+	}
+
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_resource_set_user_data(resource, NULL);
```

Comment:
This is the right thing to do, because the device resources' user data is set to the `struct wlr_drm_lease_device_v1 *`, which we're about to free in this function.